llvm/lib/Target/RISCV/RISCVFoldMasks.cpp

```diff
@@ -88,6 +98,304 @@ bool RISCVFoldMasks::isAllOnesMask(MachineInstr *MaskCopy) {
+    // FIXME: This could be relaxed, but we need to handle the policy for the
+    // resulting op correctly.
+    if (Merge->getReg() == RISCV::NoRegister)
+      return false;
+    // Both the vmerge instruction and the True instruction must have the same
+    // merge operand.
+    if (!isOpSameAs(TrueMergeOp, *False))
+      return false;
+  }
+
+  if (IsMasked) {
+    assert(HasTiedDest && "Expected tied dest");
+    // The vmerge instruction must be TU.
+    if (Merge->getReg() == RISCV::NoRegister)
+      return false;
+    // The vmerge instruction must have an all 1s mask since we're going to keep
+    // the mask from the True instruction.
+    // FIXME: Support mask agnostic True instruction which would have an
+    // undef merge operand.
+    if (BaseOpc == RISCV::VMERGE_VVM && !isAllOnesMask(MaskDef))
+      return false;
+  }
+
+  // Skip if True has side effect.
+  // TODO: Support vleff and vlsegff.
+  if (TII->get(TrueOpc).hasUnmodeledSideEffects())
+    return false;
+
+  // The vector policy operand may be present for masked intrinsics
+  const MachineOperand &TrueVL =
+      TrueMI.getOperand(RISCVII::getVLOpNum(TrueMCID));
+
+  auto GetMinVL =
+      [](const MachineOperand &LHS,
+         const MachineOperand &RHS) -> std::optional<MachineOperand> {
+    if (LHS.isReg() && RHS.isReg() && LHS.getReg().isVirtual() &&
+        LHS.getReg() == RHS.getReg())
+      return LHS;
+    if (LHS.isImm() && LHS.getImm() == RISCV::VLMaxSentinel)
+      return RHS;
+    if (RHS.isImm() && RHS.getImm() == RISCV::VLMaxSentinel)
+      return LHS;
+    if (!LHS.isImm() || !RHS.isImm())
+      return std::nullopt;
+    return LHS.getImm() <= RHS.getImm() ? LHS : RHS;
+  };
+
+  // Because MI and True must have the same merge operand (or True's operand is
+  // implicit_def), the "effective" body is the minimum of their VLs.
+  const MachineOperand VL = MI.getOperand(RISCVII::getVLOpNum(MI.getDesc()));
+  auto MinVL = GetMinVL(TrueVL, VL);
+  if (!MinVL)
+    return false;
+  bool VLChanged = !MinVL->isIdenticalTo(VL);
+
+  // If we end up changing the VL or mask of True, then we need to make sure it
+  // doesn't raise any observable fp exceptions, since changing the active
+  // elements will affect how fflags is set.
+  if (VLChanged || !IsMasked)
+    if (TrueMCID.mayRaiseFPException() &&
+        !TrueMI.getFlag(MachineInstr::MIFlag::NoFPExcept))
+      return false;
+
+  unsigned MaskedOpc = Info->MaskedPseudo;
+  const MCInstrDesc &MaskedMCID = TII->get(MaskedOpc);
+#ifndef NDEBUG
+  assert(RISCVII::hasVecPolicyOp(MaskedMCID.TSFlags) &&
+         "Expected instructions with mask have policy operand.");
+  assert(MaskedMCID.getOperandConstraint(MaskedMCID.getNumDefs(),
+                                         MCOI::TIED_TO) == 0 &&
+         "Expected instructions with mask have a tied dest.");
+#endif
+
+  // Sink True down to MI so that it can access MI's operands.
+  if (!sinkInstructionAndDeps(TrueMI, MI))
```

Comment:
I think you might be making this more complicated than it needs to be.   My mental model of this transform is that we're mutating TrueMI so as to make the merge operation a nop.  I don't think we need to (or want to) move TrueMI's operands at all.  Given the preconditions we check, I believe it is the case that False and MinVL must be operands of TrueMI.  The only remaining operand we *add* is the mask itself.  

For the mask, we have two options.
1) If the definition of V0 dominates TrueMI, then the mask is available.  In some cases, we may be able to move both the definition of the mask vreg and the copy to V0 before TrueMI.  This is a speculation problem.
2) We can sink TrueMI itself to the masked op.  We know that we *don't* need to sink any it's operands.  All we need to prove is that sinking TrueMI is sound (i.e. if TrueMI can fault in an observable way, we can't sink it below a branch or other observable fault).  Note that condition is basically the inverse of proving the mask is safe to move upwards.  (And the dual of the glue reasoning in SDAG.)

That last bit about being the dual of the glue reasoning may also be important.  The SDAG version isn't infinitely powerful, so we may not need to worry about moving around anything which "could" fault just to preserve equal power.

If I was writing this patch, I'd start by writing the simple version which tries to hoist the mask in trivial cases, and see if that was good enough.  

One thing I can't help but think - all of this would be *much* easier without the copies to V0.  Any idea why we have that?  It seems like the pseudo's could have a vreg operand with a register class which only contained V0 and have the same effect.  Why do we need to manually insert the copies so early?  I'm probably missing something, but what?  (@topperc Any context I'm missing here?)

Reply:
> I don't think we need to (or want to) move TrueMI's operands at all.

I've named this method rather misleadingly: It only sinks write-after-read dependencies of physical registers for reasons I'll explain below, not any of TrueMI's operands.

As for "hoist the mask" versus "sink TrueMI", I settled on the latter since it ended up being the simpler option.
The ordering of instructions from SelectionDAG typically looks like this:

```
%true = PsuedoVADD ...
%mask = PseudoVMSET ...
$v0 = COPY %mask
PsuedoVMERGE $v0, %true, ...
```

Where the mask is pretty much always between TrueMI and the vmerge because of the glue. So it never ends up dominating TrueMI in practice.

If we hoist the mask, the trivial case is that it's a VMSET with only some immediate operands.
However a lot of the time the mask will be computed and we'll quickly hit the wall with a naïve implementation, since there'll be a bunch of instructions that need hoisted e.g.

```
%true = PsuedoVADD ...
// All of these need hoisted
%m1 = PseudoVFOO ...
...
%mn = PseudoVBAR ...
%mask = PseudoVMSEQ %mn
$v0 = COPY %mask
PsuedoVMERGE $v0, %true, ...
```

If we sink TrueMI, there's usually much less we need to do: We check beforehand that the vmerge is the only user of TrueMI, so 99.99% of the time we can just call `TrueMI.beforeBefore(MI)`.

For loads/faulting instructions we also just need to check that there's no stores in between and `isSafeToMove` returns true. However, the annoying case is for pseudos that also read a physical register, e.g.:

```
$frm = FSRM foo
%true = PseudoVFCVT_RM_X_F ... implicit $frm
$frm = FSRM bar
%mask = PseudoVMSET ...
$v0 = COPY %mask
PsuedoVMERGE $v0, %true, ...

// can't just sink %true:

$frm = FSRM foo
$frm = FSRM bar <-- WAR dependency
%mask = PseudoVMSET ...
$v0 = COPY %mask
%true = PseudoVFCVT_RM_X_F ... implicit $frm <-- incorrect $frm
PsuedoVMERGE $v0, %true, ...
```

In which case we need to sink those write-after-read dependencies too. But this is a much rarer case in comparison to hoisting the mask with lots of dependencies, which is why I ended up settling on the sinking approach.

I am in agreement though that I would like to make this simpler though. I just want to clarify how I ended with the sinking approach. 

> One thing I can't help but think - all of this would be much easier without the copies to V0.

That thought crossed my mind as well. Is it because all the vector pseudos have their mask operands defined as `VMaskOp`?:

```tablegen
def VMaskOp : RegisterOperand<VMV0> {
  let ParserMatchClass = VMaskAsmOperand;
  let PrintMethod = "printVMaskReg";
  let EncoderMethod = "getVMaskReg";
  let DecoderMethod = "decodeVMaskReg";
}
```

Reply:
The FRM thing is something I hadn't considered.  Going to have to think about that one a bit, but I have to admit I'm very surprised to see implicit register operands for FRM this early.

Reply:
Fantastic, that should simplify things a lot.